### PR TITLE
Optionally display runahead tasks in gcylc and monitor.

### LINKED ISSUE #1073: gcylc: optionally monitor tasks in the runahead pool
Tasks sometimes temporarily disappear into the runahead pool, even in a simple suite like this:
```
[scheduling]
    initial cycle point = 20100808T06
    [[dependencies]]
        [[[T06,T18]]]
            graph = foo & bar
```
We should probably provide a switch to show them in the GUI, since this is otherwise indistinguishable from "task not present in the suite".

### PR DESCRIPTION
**Pull request by hjoliver**
Close #1073.

Tasks in the runahead pool are technically 'waiting' but inactive, and currently they are excluded from the suite state summary used in monitoring etc. If a task is present only in the runahead pool users may conclude that it has disappeared from the suite.  This branch allows optional display of runahead tasks in the GUI and terminal monitor.

Implementation:

 * include runahead tasks in the suite state summary, but convert their summary state from 'waiting' to 'runahead' (so 'runahead' becomes a task state for monitoring purposes only).
 * add a 'runahead' check-box to the GUI task state filter box
 * convert the 'unknown' task state dot icon (which there is no use for) to display runahead tasks

The runahead check-box filter is off by default because it could have a detrimental effect on graph view performance: if a suite with hourly-cycling tasks also has a yearly-cycling task waiting in the runahead pool, the graph view has to retrieve a large graph structure out to that point and then remove all the nodes in empty intermediate cycle points).

It works with all three GUI views and is backward compatible with older suite daemons - they don't have the new information in the suite state summary, so the new filter has no effect.

Unfortunately this required expanding the filter-box to three rows deep:

gcylc-runahead

### CODE REVIEW

**Comment by hjoliver** [comment]
@benfitzpatrick - please review.

**Comment by benfitzpatrick** [comment]
Maybe we should change the filters to a menu-style dropdown?

**Comment by hjoliver** [comment]
> Maybe we should change the filters to a menu-style dropdown?

I think current filter settings need to be highly visible, otherwise users forget that tasks are being filtered out ... how about this:

gcylc-filtering

The dialog box can be opened from the View menu, or by clicking on the new icons in the status bar.

**Comment by hjoliver** [comment]
(not ready for final review - I want to do a bit of tidying, and I need to figure out if an 'unknown' state icon is actually still needed)

**Comment by benfitzpatrick** [comment]
That looks great! One thing though - maybe use a different shape for 'filtered out'?

**Comment by hjoliver** [comment]
> maybe use a different shape for 'filtered out'? 

Unfortunately crosses (x) - with diagonal lines - will look pixelated as text-templated icons, and minus signs aren't very visible... any suggestions?

**Comment by hjoliver** [comment]
I've gone for simple clean exclamation marks.  Space-invader or pacman ghost icons, which are naturally pixelated, would have been nice, but they're copyrighted.

Still not ready for review, however, due to theme issues...

**Comment by hjoliver** [comment]
(I'll attempt to finish this tomorrow...)

**Comment by hjoliver** [comment]
(branch rebased)

**Comment by hjoliver** [comment]
(TODO - do we still need an 'unknown state' icon to handle bad gcylc.rc theme input, and can existing gcylc.rc files break this branch)

**Comment by hjoliver** [comment]
@benfitzpatrick - I think this is ready to go.  I've satisfied myself that older suite daemons and existing user-defined colour themes won't break this.  Note also, I've gratuitously changed the family icon indicator to "F" to make its meaning more obvious, and runahead is now treated as a normal themed task state (the fact that tasks in this state are a little different - they're held in a separate pool - is an internal implementation detail that users don't need to know).

runahead-icons

File: lib/cylc/cfgspec/gcylc.py [hunk 1]
```diff
@@ -61,7 +62,7 @@
 def upg( cfg, descr ):
     u = upgrader(cfg, descr )
     u.deprecate( '5.4.3', ['themes','__MANY__', 'submitting'], ['themes','__MANY__', 'ready'] )
-    u.deprecate( '6.0.0', ['themes','__MANY__', 'runahead'], ['themes','__MANY__', 'waiting'] )
+    #u.deprecate( '6.0.0', ['themes','__MANY__', 'runahead'], ['themes','__MANY__', 'waiting'] )
```
Comment (benfitzpatrick) [inline lib/cylc/cfgspec/gcylc.py]: We can just remove this, I guess?

File: lib/cylc/gui/app_gcylc.py [hunk 2]
```diff
@@ -107,6 +107,28 @@ def run_get_stdout(command, filter=False):
...
+    A popup window displaying task filtering options.
+    """
+    def __init__(self, parent_window, widgets, reset_task_filters):
+        super(TaskFilterWindow, self).__init__()
+        self.set_border_width(10)
+        self.set_title( "Task Filtering" )
+        if parent_window is None:
+            self.set_icon(get_icon())
+        else:
+            self.set_transient_for( parent_window )
```
Comment (benfitzpatrick) [inline lib/cylc/gui/app_gcylc.py]: Some pep8

File: lib/cylc/gui/app_gcylc.py [hunk 3]
```diff
@@ -3039,64 +3152,53 @@ def refresh_views(self):
...
+        for row in range(0, n_rows):
+            subbox = gtk.HBox(homogeneous=True)
+            self.task_filter_box.pack_start(subbox)
+            for i in range(0, PER_ROW):
+                ebox = gtk.EventBox()
+                box = gtk.HBox()
+                ebox.add(box)
+                try:
+                    st = self.legal_task_states[row * PER_ROW + i]
+                except:
```
Comment (benfitzpatrick) [inline lib/cylc/gui/app_gcylc.py]: `except Exception` is better, as `except:` catches stuff like `SystemExit` (buried under https://www.python.org/dev/peps/pep-0008/#programming-recommendations)

File: lib/cylc/gui/app_gcylc.py [hunk 4]
```diff
@@ -3039,64 +3152,53 @@ def refresh_views(self):
...
+        hbox.pack_start(ebox)
+        self.task_filter_box.pack_start(hbox)
         tooltip = gtk.Tooltips()
         tooltip.enable()
         tooltip.set_tip(self.filter_entry,
                         "Filter by task name.\n"
                         "Enter a sub-string or regex and hit Enter\n"
                         "(to reset, clear the entry and hit Enter)")
-        return padbox
+        #return padbox
```
Comment (benfitzpatrick) [inline lib/cylc/gui/app_gcylc.py]: Can be nuked

File: lib/cylc/gui/dot_maker.py [hunk 5]
```diff
@@ -462,57 +282,58 @@ def __init__(self, theme, size='medium'):
...
             else:
-                fill_color = 'None'
-                brdr_color = color
+                color = self.theme[state]['color']
+                if self.theme[state]['style'] == 'filled':
+                    filled = True
+                    fill_color = color
+                    brdr_color = color
+                else:
+                    filled = False
```
Comment (benfitzpatrick) [inline lib/cylc/gui/dot_maker.py]: `filled` doesn't do anything?

Comment (benfitzpatrick) [inline lib/cylc/gui/dot_maker.py]: Didn't read far down enough...

File: lib/cylc/gui/legend.py [hunk 6]
```diff
@@ -79,4 +79,4 @@ def _set_key_liststore(self):
         for state in task_state.legal:
             dot = dotm.get_icon(state)
             self._key_liststore.append([ state, dot])
-        self._key_liststore.append(['(unknown)', dotm.get_icon('unknown')])
+        #self._key_liststore.append(['(unknown)', dotm.get_icon('unknown')])
```
Comment (benfitzpatrick) [inline lib/cylc/gui/legend.py]: Not worth keeping?

**Comment by benfitzpatrick** [comment]
This is good stuff! A couple of minor comments.

I do think that the exclamation mark is too 'alarming', given that the default is that
runahead is filtered out. Maybe we should abandon custom images?

screenshot-mot-aa205 - gcylc

What do you think?

**Comment by benfitzpatrick** [comment]
On a separate note:
Our user feedback for 6.3.1 is that they miss the independent view filtering, and don't really like
global filtering... this change makes it much easier to re-add that facility, if we decide to do that.
Something like the default is global filtering, but clicking another button adds a separate filter for
the second view? Something for future work?

**Comment by hjoliver** [comment]
@benfitzpatrick - all points addressed.  I've gone with your suggestion on the filtered state icons.  I had originally thought that *narrow* custom icons would be better, to restrict the amount of space used, but instead of just restricted the maximum size of the standard box icons.

**Comment by hjoliver** [comment]
> Our user feedback for 6.3.1 is that they miss the independent view filtering...

They could open another GUI for that, so I don't think it's a high priority, but I'll put up an issue.

**Pull request merged**